File: python/open_guardrail/guards/knowledge_boundary.py
"""Detect when AI claims knowledge beyond its training."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_PATTERNS: list[tuple[str, re.Pattern[str]]] = [
    (
        "certain_knowledge",
        re.compile(
            r"I\s+know\s+for\s+(certain|a fact|sure)"
            r"\s+that",
            re.IGNORECASE,
        ),
    ),
    (
        "latest_data_claim",
        re.compile(
            r"the\s+latest\s+(data|statistics|numbers"
            r"|figures)\s+show",
            re.IGNORECASE,
        ),
    ),
    (
        "real_time_claim",
        re.compile(
            r"(real[- ]time|live|current)"
            r"\s+(data|price|stock|weather|status)"
            r"\s+(shows?|is|are)",
            re.IGNORECASE,
        ),
    ),
    (
        "post_cutoff_event",
        re.compile(
            r"(as\s+of|in)\s+(today|now|this\s+moment"
            r"|right\s+now),?\s+(the|we|it)",
            re.IGNORECASE,
        ),
    ),
    (
        "browsing_claim",
        re.compile(
            r"I\s+(just\s+)?(checked|browsed"
            r"|looked\s+up|searched|visited)",
            re.IGNORECASE,
        ),
    ),
    (
        "confirmed_fact",
        re.compile(
            r"I\s+can\s+confirm\s+that\s+this\s+is"
            r"\s+(true|accurate|correct)",
            re.IGNORECASE,
        ),
    ),
    (
        "definitive_future",
        re.compile(
            r"will\s+definitely"
            r"\s+(happen|occur|take\s+place)",
            re.IGNORECASE,
        ),
    ),
    (
        "absolute_claim",
        re.compile(
            r"there\s+is\s+no\s+doubt\s+that",
            re.IGNORECASE,
        ),
    ),
    (
        "current_price",
        re.compile(
            r"the\s+(current|today'?s)"
            r"\s+(price|rate|value)"
            r"\s+(of|for|is)",
            re.IGNORECASE,
        ),
    ),
    (
        "just_announced",
        re.compile(
            r"(just|recently)"
            r"\s+(announced|released|published)"
            r"\s+(today|this)",
            re.IGNORECASE,
        ),
    ),
]


class _KnowledgeBoundary:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "knowledge-boundary"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for name, pat in _PATTERNS:
            if pat.search(text):
                matched.append(name)

        triggered = len(matched) > 0
        score = (
            min(len(matched) / 3, 1.0) if triggered else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="knowledge-boundary",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Knowledge boundary violation detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched": matched} if triggered else None
            ),
        )


def knowledge_boundary(
    *, action: str = "warn"
) -> _KnowledgeBoundary:
    return _KnowledgeBoundary(action=action)

File: packages/guards/src/knowledge-boundary.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface KnowledgeBoundaryOptions {
  action: 'block' | 'warn';
}

const PATTERNS: { name: string; re: RegExp }[] = [
  { name: 'certain_knowledge', re: /I\s+know\s+for\s+(certain|a fact|sure)\s+that/i },
  { name: 'latest_data_claim', re: /the\s+latest\s+(data|statistics|numbers|figures)\s+show/i },
  { name: 'real_time_claim', re: /(real[- ]time|live|current)\s+(data|price|stock|weather|status)\s+(shows?|is|are)/i },
  { name: 'post_cutoff_event', re: /(as\s+of|in)\s+(today|now|this\s+moment|right\s+now),?\s+(the|we|it)/i },
  { name: 'browsing_claim', re: /I\s+(just\s+)?(checked|browsed|looked\s+up|searched|visited)/i },
  { name: 'confirmed_fact', re: /I\s+can\s+confirm\s+that\s+this\s+is\s+(true|accurate|correct)/i },
  { name: 'definitive_future', re: /will\s+definitely\s+(happen|occur|take\s+place)/i },
  { name: 'absolute_claim', re: /there\s+is\s+no\s+doubt\s+that/i },
  { name: 'current_price', re: /the\s+(current|today'?s)\s+(price|rate|value)\s+(of|for|is)/i },
  { name: 'just_announced', re: /(just|recently)\s+(announced|released|published)\s+(today|this)/i },
];

export function knowledgeBoundary(options: KnowledgeBoundaryOptions): Guard {
  return {
    name: 'knowledge-boundary',
    version: '0.1.0',
    description: 'Detects when AI claims knowledge beyond its training boundaries',
    category: 'ai',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const p of PATTERNS) {
        if (p.re.test(text)) {
          matched.push(p.name);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'knowledge-boundary',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched } : undefined,
      };
    },
  };
}
